### PR TITLE
Data anchor sampling.

### PR DESCRIPTION
**Pull request by qingqing01**
Fix https://github.com/PaddlePaddle/models/issues/999

### CODE REVIEW

File: fluid/face_detection/reader.py [hunk 1]
```diff
@@ -278,8 +250,8 @@ def train(settings, file_list, shuffle=True):
     return pyramidbox(settings, file_list, 'train', shuffle)
 
 
-def test(settings, file_list):
-    return pyramidbox(settings, file_list, 'test', False)
+def val(settings, file_list):
+    return pyramidbox(settings, file_list, 'val', False)
```
Comment (baiyfbupt) [inline fluid/face_detection/reader.py]: line 243 'test'->'val', same as this line

File: fluid/face_detection/image_util.py [hunk 2]
```diff
@@ -177,12 +247,41 @@ def generate_batch_samples(batch_sampler, bbox_labels, image_width,
...
+
+
+def clip_bbox(src_bbox, out_of_range=False):
+    if not out_of_range:
+        src_bbox.xmin = max(min(src_bbox.xmin, 1.0), 0.0)
+        src_bbox.ymin = max(min(src_bbox.ymin, 1.0), 0.0)
+        src_bbox.xmax = max(min(src_bbox.xmax, 1.0), 0.0)
+        src_bbox.ymax = max(min(src_bbox.ymax, 1.0), 0.0)
+        return src_bbox
+    else:
```
Comment (baiyfbupt) [inline fluid/face_detection/image_util.py]: Why this branch is needed, since it keep the same

File: fluid/face_detection/reader.py [hunk 3]
```diff
@@ -36,112 +37,84 @@ def __init__(self,
...
+            batch_sampler.append(
+                image_util.sampler(1, 50, 0.3, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0,
+                                   0.0, True))
+            batch_sampler.append(
+                image_util.sampler(1, 50, 0.3, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0,
+                                   0.0, True))
+            sampled_bbox = image_util.generate_batch_samples(
+                batch_sampler, bbox_labels, img_width, img_height)
+
+            img = np.array(img)
```
Comment (baiyfbupt) [inline fluid/face_detection/reader.py]: batch_ramdom_samples  and batch_samples all need to do `crop`?

File: fluid/face_detection/reader.py [hunk 4]
```diff
@@ -36,112 +37,84 @@ def __init__(self,
...
-    img = img.resize((settings.resize_w, settings.resize_h), Image.ANTIALIAS)
+        prob = random.uniform(0., 1.)
+        if prob > settings.data_anchor_sampling_prob:
+            scale_array = np.array([16, 32, 64, 128, 256, 512])
+            batch_sampler.append(
+                image_util.sampler(1, 10, 1.0, 1.0, 1.0, 1.0, 0.0, 0.0, 0.2,
+                                   0.0, True))
+            sampled_bbox = image_util.generate_batch_random_samples(
+                batch_sampler, bbox_labels, img_width, img_height, scale_array,
+                settings.resize_width, settings.resize_height)
```
Comment (baiyfbupt) [inline fluid/face_detection/reader.py:87]: need to do `crop_image_sampling` in random_samples case

File: fluid/face_detection/reader.py (hunk 1 above)
Comment (qingqing01) [reply fluid/face_detection/reader.py]: Done.

File: fluid/face_detection/reader.py (hunk 4 above)
Comment (qingqing01) [reply fluid/face_detection/reader.py:87]: Fixed. Thanks!

File: fluid/face_detection/image_util.py (hunk 2 above)
Comment (qingqing01) [reply fluid/face_detection/image_util.py]: remote this branch.

File: fluid/face_detection/image_util.py [hunk 5]
```diff
@@ -177,6 +247,28 @@ def generate_batch_samples(batch_sampler, bbox_labels, image_width,
     return sampled_bbox
 
 
+def generate_batch_random_samples(batch_sampler, bbox_labels, image_width,
+                                  image_height, scale_array, resize_width,
+                                  resize_height):
+    sampled_bbox = []
+    index = []
+    c = 0
```
Comment (baiyfbupt) [inline fluid/face_detection/image_util.py]: is index and c needed? seems never be used

Comment (qingqing01) [reply fluid/face_detection/image_util.py]: Removed.

**Review by baiyfbupt — APPROVED** [review]
LGTM

**Pull request merged**